MudDataGrid: Add extensions method on `IQueryable` for easy filtering and ordering

0xced opened the pull request:
## Description

This pull request introduces two new extensions methods on `IQueryable`: `Where` that takes a `IEnumerable<IFilterDefinition<T>>` and `OrderBy` that takes a `IEnumerable<SortDefinition<T>>`.

This enables a straightforward implementations of the `ServerData` function, for example with EF Core.
```csharp
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

namespace MudBlazor;

internal static class QueryGridExtensions
{
    public static async Task<GridData<T>> LoadServerDataAsync<T>(this IQueryable<T> source, GridState<T> state, CancellationToken cancellationToken = default)
    {
        var query = source.Where(state.FilterDefinitions).OrderBy(state.SortDefinitions);
        var items = await query.Skip(state.Page * state.PageSize).Take(state.PageSize).ToListAsync(cancellationToken);
        var totalItems = await query.CountAsync(cancellationToken);
        return new GridData<T> { Items = items, TotalItems = totalItems };
    }
}
```

(A good implementation would need to order the source with a default ordering in case the sort definitions are empty.)

It also fixes #6682.

## How Has This Been Tested?
Two new unit tests have been added: `QueryFilterExtensionTest` and `QuerySortExtensionTest`.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

vernou commented on src/MudBlazor/Components/DataGrid/QueryFilterExtensions.cs:
Maybe NRT can be enable on this file? Like you did in **QuerySortExtensions.cs**.

0xced replied:
Good point! I did in in f60b02125909eabb9759523f35ca9779890d11a5, rebased on the `dev` branch and force-pushed.

Militancy commented:
Hey @0xced,

thanks for your pull request, I currently struggled to find a way to do those queries user friendly. For myself I would like that extension to be a bit more powerful.

Would you consider to implement an solution where its possible to query for included class structures, like for example extend the MemberExpression for each member of depth?

vernou commented:
@Militancy 
> Would you consider to implement an solution where its possible to query for included class structures, like for example extend the MemberExpression for each member of depth?

I am not sure I understand...
Have you a example?
Will the resulting expressions be EF compatible?

Militancy commented:
@vernou 
In a specific example we have two classes, where we want to query for a value of a nested class. As i tested your stuff i noticed that the the extension for filtering was already working for nested class properties. Only on the sorting part a adjustment was necessary.

I tested it with my own class structures with EF on SQLite successfully.

Example Column for MudDataGrid
`<PropertyColumn Property="p => p.B.Name"/>`
 
Class example
```
public class A {
    public int BId { get; set; }
    public B { get; set; }
}

public class B {
    public string Name { get; set; }
}
```


A "solution" would be sth like this. But I dont know if its the best way to go.

```
MemberExpression propertyExpression;
if(sortDefinition.SortBy.Contains('.'))
{
    var properties = sortDefinition.SortBy.Split('.');
    propertyExpression = Expression.Property(parameter, properties.First());
    foreach(var prop in properties.Skip(1))
        propertyExpression = Expression.Property(propertyExpression, prop);
}
else
    propertyExpression = Expression.Property(parameter, sortDefinition.SortBy);
```

0xced commented:
@Militancy I think I figured out a way to support your scenario. Instead of adding one extension method, I added two extension methods (and force pushed my branch). The new one that I added takes an additional `Func<ParameterExpression, SortDefinition<T>, Expression>` that lets you define the expression as you see fit for the given sort definition. The default implementation for the standard case does `Expression.Property(parameter, sortDefinition.SortBy)` but now you can write your own if that's not enough.

@ScarletKuro I don't think that's a problem since the type `T` is user supplied and thus very unlikely to be trimmed.

0xced committed:
MudDataGrid: Add extensions method on IQueryable for easy sorting

Fixes #6682

0xced committed:
MudDataGrid: Add an extension method on IQueryable for easy filtering

0xced commented:
Hey @tjscience, could you please have a look at this pull request when you get a chance?

tjscience reviewed (approved):
I don't see any issues with this given that they are just extension methods.

ScarletKuro commented:
@0xced Can this two https://github.com/MudBlazor/MudBlazor/pull/9181 https://github.com/MudBlazor/MudBlazor/pull/9183 be closed then?
